enhancements/aws-permissions-check-bypass.md

```diff
@@ -0,0 +1,269 @@
+OpenShift releases.
+
+In-cluster users of cloud credentials (image-registry, ingress-operator,
+machine-api-operator) will be exposed to situations where the credentials that
+were requested via the CredentialsRequest CRs have not been validated in any way
+when CCO bypasses the permissions verficiation before handing over credentials
+to satisfy a CredentialsRequest.
+
+## Design Details
+
+### Test Plan
+
+Ideally, e2e coverage of installing OpenShift in an AWS account with SCP
+permissions defined in a way that would otherwise fail without bypassing
+permissions checking. Acceptably, simply running the installation with the
+environment variable defined to force bypassing permissions checking.
+
+We should also consider the case where OpenShift was installed in an AWS account
+without SCPs defined, but the account is then migrated to an environment where
+SCPs are now defined post-installation. CCO should be able to recover by an
+admin defining the ConfigMap to force CCO into `mint` or `passthrough` mode as
+appropriate. CCO will eventually settle into a functioning state (assuming the
+credentials have sufficient permissions).
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. Initial proposal
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this
+enhancement:
+- Maturity levels - `Dev Preview`, `Tech Preview`, `GA`
+- Deprecation
+
+Clearly define what graduation means.
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned
+[maturity levels][maturity-levels].
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA 
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+##### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+N/A. A running cluster whose AWS account is migrated to an organization that
+does define SCPs will start to error.
+
+During cluster runtime, enabling/disabling the bypassing of permissions
+simulations can be controlled through the contents of the ConfigMap.
+
+### Version Skew Strategy
+
+The installer generates the configmap for CCO, and it will have to stay in sync
+with the format for delivering this `mint` vs `passthrough` information if/when
+CCO moves away from using a ConfigMap.
+
+## Implementation History
+
```

Comment:
i think we should include previous efforts of improving the simulation, like the aws region constraint etc.

Reply:
I checked, and it seems like a limitation with AWS. As a concrete example, I created an SCP with the limitation that `rds:DescribeDBInstances` is only allowed if the region is `us-east-1`:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Statement1",
      "Effect": "Deny",
      "Action": [
        "rds:DescribeDBInstances"
      ],
      "Resource": [
        "*"
      ],
      "Condition": {
        "StringNotEquals": {
          "aws:RequestedRegion": [
            "us-east-1"
          ]
        }
      }
    }
  ]
}
```

And the user can make the API call, but simulating while providing the region info still indicates that it would have failed.

```
[jdiaz@minigoomba ~]$ aws rds describe-db-instances --region us-east-1
{
    "DBInstances": []
}

[jdiaz@minigoomba ~]$ aws rds describe-db-instances --region us-east-2

An error occurred (AccessDenied) when calling the DescribeDBInstances operation: User: arn:aws:iam::477645798544:user/jdiaz is not authorized to perform: rds:DescribeDBInstances with an explicit deny

[jdiaz@minigoomba ~]$ aws iam simulate-principal-policy --policy-source-arn arn:aws:iam::ACCOUNT_NUMBER:user/jdiaz --action-names '["rds:DescribeDBInstances"]' --context-entries ContextKeyName=aws:RequestedRegion,ContextKeyType=string,ContextKeyValues=us-east-1 | jq -r .EvaluationResults[0].EvalDecision
implicitDeny
```

Meanwhile, the exact same test when the restriction on the API call is defined as a permissions boundary on the user gives accurate permissions simulation results.